src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,216 @@
+            if (argument != null)
+            {
+                conversionType = null;
+                var argumentListNode = argument.Parent;
+                var invocationNode = argumentListNode.Parent;
+                var symbolInfo = semanticModel.GetSymbolInfo(invocationNode, cancellationToken);
+                var candidcateSymbols = symbolInfo.CandidateSymbols;
+
+                var argumentList = (ArgumentListSyntax)argumentListNode;
+
+                // Find all valid candidates and extract potential conversion types 
+                var potentialConversionTypes = new List<ITypeSymbol> {};
+                foreach (var candidcateSymbol in candidcateSymbols)
+                {
+                    var methodSymbol = (IMethodSymbol)candidcateSymbol;
+                    var parameterList = methodSymbol.Parameters;
+                    if (parameterList.Length != argumentList.Arguments.Count)
+                    {
+                        continue;
+                    }
+
+                    // Test if all parameters are convertible, otherwise it is not the perfect match function
+                    var allValid = true;
+                    for (var i = 0; i < parameterList.Length; i++)
+                    {
+                        var argType = semanticModel.GetTypeInfo(argumentList.Arguments[i].Expression, cancellationToken);
+                        if (!IsTypeConvertible(argType.Type, parameterList[i].Type))
+                        {
+                            allValid = false;
+                            break;
+                        }
+                    }
+
+                    if (allValid)
+                    {
+                        var targetArgumentIndex = argumentList.Arguments.IndexOf((ArgumentSyntax)argument);
+                        var correspondingParameter = parameterList[targetArgumentIndex];
+                        potentialConversionTypes.Add(correspondingParameter.Type);
+                    }
+                }
+
+                // If there is no exact solution, then don't provide suggestions
+                if (potentialConversionTypes.Count != 1)
+                {
+                    return;
+                }
+
+                conversionType = potentialConversionTypes[0];
+            }
+        }
+
+        private bool IsTypeConvertible(ITypeSymbol nodeType, ITypeSymbol conversionType)
+        {
+            if (conversionType == null) return false;
+            if (nodeType.Equals(conversionType)) return true;
+
+            var convertible = false;
+            if (conversionType.Interfaces.Length != 0)
+            {
+                foreach(var interface_type in conversionType.Interfaces)
+                {
+                    convertible |= IsTypeConvertible(nodeType, interface_type);
+                }
+            }
+            convertible |= IsTypeConvertible(nodeType, conversionType.BaseType);
+
+            return convertible;
```

Comment:
i don't understand what is going on with this functin.  can you provide docs that explain what the intuition around it is?